main.tf

```diff
@@ -1,6 +1,6 @@
 locals {
   default_resource_group_name = "${var.product}-${var.env}${var.deployment_target}"
-  resource_group_name         = "${var.resource_group_name != "" ? var.resource_group_name : local.default_resource_group_name}"
+  resource_group_name         = "${coalesce(var.resource_group_name, local.default_resource_group_name)}"
```

Comment:
this change should be fine to be mainlined if you separate it out to a different PR

Reply:
If the intent is to declare a Resource Group externally to the module and pass it in as a module parameter then this change is can be removed in favour of:
```suggestion
```

Reply:
unfortunately there's backwards compatibility to keep and we can't really make that change as it would be breaking

Reply:
@timja I think it's not really achievable to resolve these execution issues, support an external resource group declaration and retain backward compatibility.

Terraform 0.11.x doesn't really have sufficient expressions and logic to do this and obviously 0.12.x is a breaking change in itself.

I have tried a few theories and below are my findings.

Use cases:
- ⚠ cnp-module-webapps defined without `asp_rg` (No external RG)
```
A conditional resource in the module will create a resource group with default name.
NOTE: This does NOT run first time. Due to limitations in Terraform 0.11.x this implementation will work but will first fail with a 404 ResourceGroupNotFound error.
```
- ❗️cnp-module-webapps defined with `asp_rg` (No external RG)
```
The module will assume that the resource group exists and will use `local.asp_rg` for all resource group references. 
NOTE: This will fail if the resource group does not exist!
```
- ❗️cnp-module-webapps defined with `asp_rg` and resource group exists already due to a previous execution (No external RG)
```
The module will plan destruction of the resource group and cause the module to fail.
```
- ✅ cnp-module-webapps defined with `asp_rg = ${azurerm_resource_group.default.name}` (External RG)
```
The module will plan destruction of the resource group and cause the module to fail.
```

Due to this and the planned deprecation of this module, I'm going to have to suggest that we close this PR for good and submit a PR for allowing the use of this branch in the Terraform whitelist.  

Do you agree?

Reply:
Somewhat, but when we move to an external resource group we do a state migration (https://github.com/hmcts/terraform-state-move)